src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+                switch (const auto tag = DWRITE_MAKE_FONT_AXIS_TAG(s[0], s[1], s[2], s[3]))
+                {
+                case DWRITE_FONT_AXIS_TAG_WEIGHT:
+                    fontAxisValues[0].value = p.second;
+                    break;
+                case DWRITE_FONT_AXIS_TAG_ITALIC:
+                    fontAxisValues[1].value = p.second;
+                    break;
+                case DWRITE_FONT_AXIS_TAG_SLANT:
+                    fontAxisValues[2].value = p.second;
+                    break;
+                default:
+                    fontAxisValues.emplace_back(DWRITE_FONT_AXIS_VALUE{ tag, p.second });
+                    break;
+                }
+            }
+        }
+    }
+
+    const auto previousCellSize = _api.fontMetrics.cellSize;
+    _resolveFontMetrics(fontInfoDesired, fontInfo, &_api.fontMetrics);
+    _api.fontFeatures = std::move(fontFeatures);
+    _api.fontAxisValues = std::move(fontAxisValues);
+
+    WI_SetFlag(_api.invalidations, ApiInvalidations::Font);
+
+    if (previousCellSize != _api.fontMetrics.cellSize)
+    {
+        _api.cellCount = _api.sizeInPixel / _api.fontMetrics.cellSize;
+        WI_SetFlag(_api.invalidations, ApiInvalidations::Size);
+    }
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+void AtlasEngine::UpdateHyperlinkHoveredId(const uint16_t hoveredId) noexcept
+{
+}
+
+#pragma endregion
+
+void AtlasEngine::_resolveFontMetrics(const FontInfoDesired& fontInfoDesired, FontInfo& fontInfo, FontMetrics* fontMetrics) const
+{
+    auto requestedFaceName = fontInfoDesired.GetFaceName().c_str();
+    const auto requestedFamily = fontInfoDesired.GetFamily();
+    auto requestedWeight = fontInfoDesired.GetWeight();
+    auto requestedSize = fontInfoDesired.GetEngineSize();
+
+    if (!requestedFaceName)
+    {
+        requestedFaceName = L"Consolas";
+    }
+    if (!requestedSize.Y)
+    {
+        requestedSize = { 0, 12 };
+    }
+    if (!requestedWeight)
+    {
+        requestedWeight = DWRITE_FONT_WEIGHT_NORMAL;
+    }
+
+    wil::com_ptr<IDWriteFontCollection> systemFontCollection;
+    THROW_IF_FAILED(_sr.dwriteFactory->GetSystemFontCollection(systemFontCollection.addressof(), false));
+
+    u32 index = 0;
+    BOOL exists = false;
+    THROW_IF_FAILED(systemFontCollection->FindFamilyName(requestedFaceName, &index, &exists));
+    THROW_HR_IF(DWRITE_E_NOFONT, !exists);
+
+    wil::com_ptr<IDWriteFontFamily> fontFamily;
+    THROW_IF_FAILED(systemFontCollection->GetFontFamily(index, fontFamily.addressof()));
+
+    wil::com_ptr<IDWriteFont> font;
+    THROW_IF_FAILED(fontFamily->GetFirstMatchingFont(static_cast<DWRITE_FONT_WEIGHT>(requestedWeight), DWRITE_FONT_STRETCH_NORMAL, DWRITE_FONT_STYLE_NORMAL, font.addressof()));
+
+    wil::com_ptr<IDWriteFontFace> fontFace;
+    THROW_IF_FAILED(font->CreateFontFace(fontFace.addressof()));
+
+    DWRITE_FONT_METRICS metrics;
+    fontFace->GetMetrics(&metrics);
+
+    static constexpr u32 codePoint = L'M';

```

Comment:
> L'M'; [](http://example.com/codeflow?start=36&length=6)

should this be shared with the spot above where you also define M?

Reply:
Nah I think it's fine if they go their separate ways, so long as the GDI code is commented out...